File: backend/api/assets/starters/python-linkedlist/submission/LinkedList.py
import java.util.*;

public class LinkedList {
    private static class Node {
        int value;
        Node next;
        Node(int v) { value = v; }
    }

    public static class IntBox { public int value; }

    private Node head, tail;
    private int size;

    public LinkedList() {}

    // ── INTENTIONAL BUG: shallow copy (alias nodes instead of cloning) ──
    public LinkedList(LinkedList other) {
        // alias all internal state rather than copying
        this.head = other.head;
        this.tail = other.tail;
        this.size = other.size;
    }

    // move-like (static)
    public static LinkedList moveFrom(LinkedList src) {
        LinkedList dst = new LinkedList();
        dst.head = src.head;
        dst.tail = src.tail;
        dst.size = src.size;
        src.head = src.tail = null;
        src.size = 0;
        return dst;
    }

    // move-assign-like
    public void moveAssignFrom(LinkedList src) {
        if (this == src) return;
        clear();
        this.head = src.head;
        this.tail = src.tail;
        this.size = src.size;
        src.head = src.tail = null;
        src.size = 0;
    }

    public boolean isEmpty() { return size == 0; }
    public boolean empty() { return isEmpty(); }
    public int size() { return size; }

    public void clear() {
        Node n = head;
        while (n != null) {
            Node next = n.next;
            n.next = null;
            n = next;
        }
        head = tail = null;
        size = 0;
    }

    public void pushFront(int v) {
        Node n = new Node(v);
        n.next = head;
        head = n;
        if (tail == null) tail = n;
        size++;
    }

    public void pushBack(int v) {
        Node n = new Node(v);
        if (tail != null) {
            tail.next = n;
            tail = n;
        } else {
            head = tail = n;
        }
        size++;
    }

    public boolean popFront(IntBox out) {
        if (head == null) return false;
        if (out != null) out.value = head.value;
        head = head.next;
        // INTENTIONAL BUG: if list becomes empty, tail should be null (not done)
        size--;
        return true;
    }

    public int front() {
        if (head == null) throw new IllegalStateException("empty");
        return head.value;
    }

    public int back() {
        if (tail == null) throw new IllegalStateException("empty");
        return tail.value;
    }

    public boolean insert(int index, int value) {
        if (index < 0 || index > size) return false;
        if (index == 0) { pushFront(value); return true; }
        if (index == size) { pushBack(value); return true; }
        Node prev = head;
        for (int i = 0; i + 1 < index; i++) prev = prev.next;
        Node n = new Node(value);
        n.next = prev.next;
        prev.next = n;
        size++;
        return true;
    }

    public boolean erase(int index) {
        if (index < 0 || index >= size) return false;
        if (index == 0) {
            return popFront(new IntBox()); // inherits tail bug
        }
        Node prev = head;
        for (int i = 0; i + 1 < index; i++) prev = prev.next;
        Node victim = prev.next;
        prev.next = victim.next;
        // INTENTIONAL BUG: if removing last element, tail should become prev (not done)
        size--;
        return true;
    }

    public List<Integer> toList() {
        ArrayList<Integer> out = new ArrayList<>(size);
        for (Node n = head; n != null; n = n.next) out.add(n.value);
        return out;
    }

    // ── also make copyFrom shallow to be consistent (not used by lecturer Main, but keeps story straight) ──
    private void copyFrom(LinkedList other) {
        this.head = other.head;
        this.tail = other.tail;
        this.size = other.size;
    }
}

File: backend/api/assets/starters/go-linkedlist/memo/linked_list.go
package main

type node struct {
    val  int
    next *node
}

type LinkedList struct {
    head *node
    tail *node
    size int
}

func New() *LinkedList { return &LinkedList{} }
func (l *LinkedList) Len() int { return l.size }
func (l *LinkedList) IsEmpty() bool { return l.size == 0 }

func (l *LinkedList) Clear() {
    for l.head != nil {
        n := l.head
        l.head = n.next
        n.next = nil
    }
    l.tail = nil
    l.size = 0
}

func (l *LinkedList) PushFront(v int) {
    n := &node{val: v, next: l.head}
    l.head = n
    if l.tail == nil { l.tail = n }
    l.size++
}

func (l *LinkedList) PushBack(v int) {
    n := &node{val: v}
    if l.tail == nil { l.head, l.tail = n, n } else { l.tail.next = n; l.tail = n }
    l.size++
}

func (l *LinkedList) PopFront() (bool, int) {
    if l.head == nil { return false, 0 }
    n := l.head
    l.head = n.next
    if l.head == nil { l.tail = nil }
    l.size--
    return true, n.val
}

func (l *LinkedList) Front() (int, bool) {
    if l.head == nil { return 0, false }
    return l.head.val, true
}

func (l *LinkedList) Back() (int, bool) {
    if l.tail == nil { return 0, false }
    return l.tail.val, true
}

func (l *LinkedList) InsertAt(idx int, v int) bool {
    if idx < 0 || idx > l.size { return false }
    if idx == 0 { l.PushFront(v); return true }
    if idx == l.size { l.PushBack(v); return true }
    prev := l.head
    for i := 0; i < idx-1; i++ { prev = prev.next }
    n := &node{val: v, next: prev.next}
    prev.next = n
    l.size++
    return true
}

func (l *LinkedList) RemoveAt(idx int) bool {
    if idx < 0 || idx >= l.size { return false }
    if idx == 0 {
        ok, _ := l.PopFront(); return ok
    }
    prev := l.head
    for i := 0; i < idx-1; i++ { prev = prev.next }
    victim := prev.next
    prev.next = victim.next
    if victim == l.tail { l.tail = prev }
    l.size--
    return true
}

func (l *LinkedList) ToSlice() []int {
    out := make([]int, 0, l.size)
    for n := l.head; n != nil; n = n.next { out = append(out, n.val) }
    return out
}

func (l *LinkedList) Copy() *LinkedList {
    dst := New()
    for n := l.head; n != nil; n = n.next { dst.PushBack(n.val) }
    return dst
}

func MoveFrom(src *LinkedList) *LinkedList {
    dst := New()
    dst.head, dst.tail, dst.size = src.head, src.tail, src.size
    src.head, src.tail, src.size = nil, nil, 0
    return dst
}

func (l *LinkedList) MoveAssignFrom(src *LinkedList) {
    l.Clear()
    l.head, l.tail, l.size = src.head, src.tail, src.size
    src.head, src.tail, src.size = nil, nil, 0
}

